frigate/events/audio.py

```diff
@@ -0,0 +1,220 @@
+    setproctitle("frigate.audio_manager")
+    listen()
+
+    for camera in config.cameras.values():
+        if camera.enabled and camera.audio.enabled:
+            audio = AudioEventMaintainer(camera, stop_event)
+            audio_threads.append(audio)
+            audio.start()
+
+
+class AudioTfl:
+    def __init__(self):
+        self.labels = load_labels("/audio-labelmap.txt")
+        self.interpreter = Interpreter(
+            model_path="/cpu_audio_model.tflite",
+            num_threads=2,
+        )
+
+        self.interpreter.allocate_tensors()
+
+        self.tensor_input_details = self.interpreter.get_input_details()
+        self.tensor_output_details = self.interpreter.get_output_details()
+
+    def _detect_raw(self, tensor_input):
+        self.interpreter.set_tensor(self.tensor_input_details[0]["index"], tensor_input)
+        self.interpreter.invoke()
+        detections = np.zeros((20, 6), np.float32)
+
+        res = self.interpreter.get_tensor(self.tensor_output_details[0]["index"])[0]
+        non_zero_indices = res > 0
+        class_ids = np.argpartition(-res, 20)[:20]
+        class_ids = class_ids[np.argsort(-res[class_ids])]
+        class_ids = class_ids[non_zero_indices[class_ids]]
+        scores = res[class_ids]
+        boxes = np.full((scores.shape[0], 4), -1, np.float32)
+        count = len(scores)
+
+        for i in range(count):
+            if scores[i] < 0.4 or i == 20:
+                break
+            detections[i] = [
+                class_ids[i],
+                float(scores[i]),
+                boxes[i][0],
+                boxes[i][1],
+                boxes[i][2],
+                boxes[i][3],
+            ]
+
+        return detections
+
+    def detect(self, tensor_input, threshold=0.8):
+        detections = []
+
+        raw_detections = self._detect_raw(tensor_input)
+
+        for d in raw_detections:
+            if d[1] < threshold:
+                break
+            detections.append(
+                (self.labels[int(d[0])], float(d[1]), (d[2], d[3], d[4], d[5]))
+            )
+        return detections
+
+
+class AudioEventMaintainer(threading.Thread):
+    def __init__(self, camera: CameraConfig, stop_event: mp.Event) -> None:
+        threading.Thread.__init__(self)
+        self.name = f"{camera.name}_audio_event_processor"
+        self.config = camera
+        self.detections: dict[dict[str, any]] = {}
+        self.stop_event = stop_event
+        self.detector = AudioTfl()
+        self.shape = (int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE)),)
+        self.chunk_size = int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE * 2))
+        self.pipe = f"{CACHE_DIR}/{self.config.name}-audio"
+        self.ffmpeg_cmd = get_ffmpeg_arg_list(
+            FFMPEG_COMMAND.format(
+                " ".join(
+                    self.config.ffmpeg.global_args
+                    + parse_preset_input("preset-rtsp-audio-only", 1)
+                ),
```

Comment:
```
            FFMPEG_COMMAND.format(
                " ".join(self.config.ffmpeg.global_args)
                + " ".join(parse_preset_input("preset-rtsp-audio-only", 1)),
```

Reply:
that won't work, there won't be a space between the end of the first and the beginning of the second groups. Also in general I see nothing wrong with the way it is.

```
>>> print(" ".join(["one", "two"]) + " ".join(["three", "four"]))
one twothree four
```

Reply:
yeah, it's my copy-paste mistake

```
    def __init__(self, camera: CameraConfig, stop_event: mp.Event) -> None:
        threading.Thread.__init__(self)
        self.name = f"{camera.name}_audio_event_processor"
        self.config = camera
        self.detections: dict[dict[str, any]] = {}
        self.stop_event = stop_event
        self.detector = AudioTfl()
        self.shape = (int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE)),)
        self.chunk_size = int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE * 2))
        self.pipe = f"{CACHE_DIR}/{self.config.name}-audio"
        self.ffmpeg_cmd = get_ffmpeg_arg_list(
            self.get_ffmpeg_audio_command("preset-rtsp-audio-only")
        )
        self.pipe_file = None
        self.logpipe = LogPipe(f"ffmpeg.{self.config.name}.audio")
        self.audio_listener = None

    def get_ffmpeg_audio_command(self, preset: str) -> list:
        global_args = " ".join(self.config.ffmpeg.global_args)
        preset_arts = " ".join(parse_preset_input(preset))

        return f"ffmpeg {{}} -i {{}} -f {AUDIO_FORMAT} -ar {AUDIO_SAMPLE_RATE} -ac 1 -y {{}}".format(
            global_args + preset_arts,
            ([i.path for i in self.config.ffmpeg.inputs if "audio" in i.roles][0]),
            self.pipe,
        )
````
(commit https://github.com/skrashevich/frigate/commit/38c1c8fdb4b986ee6a49ec293ff51085c6f8deab)

Reply:
your version don't work:
```
2023-06-20 10:39:16.347863469  Process audio_capture:
2023-06-20 10:39:16.356396537  Traceback (most recent call last):
2023-06-20 10:39:16.356405545    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2023-06-20 10:39:16.356407064      self.run()
2023-06-20 10:39:16.356408104    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2023-06-20 10:39:16.356414650      self._target(*self._args, **self._kwargs)
2023-06-20 10:39:16.356415962    File "/opt/frigate/frigate/events/audio.py", line 62, in listen_to_audio
2023-06-20 10:39:16.356418970      audio = AudioEventMaintainer(camera, process_info, stop_event)
2023-06-20 10:39:16.356444799    File "/opt/frigate/frigate/events/audio.py", line 154, in __init__
2023-06-20 10:39:16.356446205      self.config.ffmpeg.global_args
2023-06-20 10:39:16.356447054  TypeError: can only concatenate str (not "list") to str
2023-06-20 10:39:16.401458498  10:39:16.401 DBG [rtsp] new consumer stream=livingroom
2023-06-20 10:39:16.401499738  10:39:16.401 DBG [rtsp] new consumer stream=kitchen
2023-06-20 10:39:16.417735036  10:39:16.417 DBG [streams] start producer url=rtsp://19
```

Reply:
It does work, I was literally testing audio events with that code for quite a while. Unless there is some platform difference or weird bug with it. I'll take a look.

Reply:
> It does work, I was literally testing audio events with that code for quite a while. Unless there is some platform difference or weird bug with it. I'll take a look.

Show the minimum example of the working config

Reply:
I see, using the default config it works but if global_args are defined then it is a string so it doesn't work.